inference: enable constant propagation for union-split signatures

aviatesk opened the pull request:
The inference precision of certain functions really relies on constant
propagation, but currently constant prop' won't happen when a call
signature is union split and so sometimes inference ends up looser
return type: e.g.
```julia
julia>
Base.return_types((Union{Tuple{Int,Nothing},Tuple{Int,Missing}},)) do t
           a, b = t
           a # I expected a::Int, but a::Union{Missing,Nothing,Int}
       end |> first Union{Missing, Nothing, Int64}
```

This PR:
- enables constant prop' for each union signatures, by calling
  `abstract_call_method_with_const_args` just after each
  `abstract_call_method`
- refactor `abstract_call_method_with_const_args` into two separate 
  parts, 1.) heuristics to decide whether to do constant prop', 2.) try
  constant propagation

The added test cases will should showcase the cases where the
inference result could be improved by that.

---

I've not seen notable regression in latency with this PR.
Here is a sample benchmark of the impact of this PR on latency,
from which I guess this PR is acceptable ?

> build time: master (caeacef)
```bash
Sysimage built. Summary:
Total ───────  61.615938 seconds
Base: ───────  26.575732 seconds 43.1313%
Stdlibs: ────  35.038024 seconds 56.8652%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 30/30
Executing precompile statements... 1378/1378
Precompilation complete. Summary:
Total ─────── 116.417013 seconds
Generation ──  81.077365 seconds 69.6439%
Execution ───  35.339648 seconds 30.3561%
    LINK usr/lib/julia/sys.dylib
```

> build time: this PR
```bash
Stdlibs total  ──── 34.077962 seconds
Sysimage built. Summary:
Total ───────  61.804573 seconds
Base: ───────  27.724077 seconds 44.8576%
Stdlibs: ────  34.077962 seconds 55.1383%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 30/30
Executing precompile statements... 1362/1362
Precompilation complete. Summary:
Total ─────── 111.262672 seconds
Generation ──  83.535305 seconds 75.0794%
Execution ───  27.727367 seconds 24.9206%
    LINK usr/lib/julia/sys.dylib
```

> first time to plot: master (caeacef)
```julia
julia> using Plots; @time plot(rand(10,3))
  3.614168 seconds (5.47 M allocations: 324.564 MiB, 5.73% gc time,
53.02% compilation time)
```

> first time to plot: this PR
```julia
julia> using Plots; @time plot(rand(10,3))
  3.557919 seconds (5.53 M allocations: 328.812 MiB, 2.89% gc time,
51.94% compilation time)
```

---

- fixes #37610
- some part of this code was taken from #37637
- this PR is originally supposed to be alternative and more generalized
  version of #39296

KristofferC commented:
Related to https://github.com/JuliaLang/julia/pull/37637? (Also ref https://github.com/JuliaLang/julia/issues/37610).

aviatesk commented:
> Related to #37637? (Also ref #37610).

Ah yeah, I haven't noticed there's already an attempt.

Can I take the initiative on this ? I agree with [@vtjnash 's comment](https://github.com/JuliaLang/julia/pull/37637#pullrequestreview-491618899), and I'd like to refactor `abstract_call_method_with_const_args` so that we can make more sensible decision earlier about whether to do constant prop' before we ends up `Any` return result on 2nd or subsequent constant prop' trial.

JeffBezanson commented:
Thanks for working on this. I think we should refactor this code even more. Specifically, I would try to use a single loop over the applicable methods, and for each signature also try constant propagation if it might be profitable --- i.e. add a call to `abstract_call_method_with_const_args` for each call to `abstract_call_method` (with appropriate guards to skip it when possible).

aviatesk commented:
Okay, I think this is in good shape. But I'd like to wait for #38905 and #39116 to be merged before asking your reviews, because both of them include certain amount of changes around `abstract_call_gf_by_type`

JeffBezanson commented on test/compiler/inference.jl:
```diff
@@ -3008,3 +3013,65 @@ f38888() = S38888(Base.inferencebarrier(3))
+        b
+    end == Any[Union{String,Char}]
+
+    @test (@evaltoplevel begin
```
Can this just use `@eval` instead?

aviatesk replied:
Does this look better ?
```julia
    anonymous_module() = Core.eval(@__MODULE__, :(module $(gensym()) end))::Module
    ...
    m = anonymous_module()
    @test (@eval m begin
        struct F32
            val::Float32
            _v::Int
        end
        struct F64
            val::Float64
            _v::Int
        end
        Base.return_types((Union{F32,F64},)) do f
            f.val
        end
    end) == Any[Union{Float32,Float64}]
```

JeffBezanson commented:
Very nice! I think this is basically the right design.

aviatesk commented:
Merged #38905 into this branch to see if it works w/o problem. 

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/39305#issuecomment-774411532):
```
ProcessExitedException(2)
```
cc @christopher-dG

aviatesk commented:
I'd like to push this forward without waiting for #38905 to be merged; this could be reviewed as is.
/cc @JeffBezanson @Keno @vtjnash

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -519,6 +374,182 @@ function abstract_call_method(interp::AbstractInterpreter, method::Method, @nosp
+function maybe_get_const_prop_profitable(interp::AbstractInterpreter, @nospecialize(rettype),
+                                         @nospecialize(f), argtypes::Vector{Any}, match::MethodMatch,
+                                         sv::InferenceState, edgecycle::Bool)
+    const_prop_entry_heuristic(interp, rettype, sv, edgecycle) || return nothing
+    method = match.method
+    nargs::Int = method.nargs
+    method.isva && (nargs -= 1)
+    length(argtypes) >= nargs || return nothing
+    const_prop_argument_heuristic(interp, argtypes) || const_prop_rettype_heuristic(interp, rettype) || return nothing
+    allconst = is_allconst(argtypes)
+    force = force_const_prop(interp, f, method)
+    force || const_prop_function_heuristic(interp, f, argtypes, nargs, allconst) || return nothing
+    force |= allconst
+    mi = specialize_method(match, !force)
+    mi === nothing && return nothing
+    mi = mi::MethodInstance
+    force || const_prop_methodinstance_heuristic(interp, method, mi) || return nothing
+    return mi
```
@vtjnash What do you think on these `xxx_heuristic` functions ? I intentionally parameterized so that it'd allow us to tweak constant prop' heuristics more easier, but I wonder the entire constant prop' heuristic is now harder to grasp as a whole. Should we inline them into `maybe_get_const_prop_profitable` ? (thinking https://github.com/JuliaLang/julia/pull/39439#discussion_r575565179)

vtjnash replied:
Makes sense to split these out and give them separate names, even though it does break up the compact operation of the function

aviatesk committed:
inference: enable constant propagation for union-split signatures

The inference precision of certain functions really relies on constant
propagation, but currently constant prop' won't happen when a call
signature is union split and so sometimes inference ends up looser
return type: e.g.
```julia
julia>
Base.return_types((Union{Tuple{Int,Nothing},Tuple{Int,Missing}},)) do t
           a, b = t
           a # I expected a::Int, but a::Union{Missing,Nothing,Int}
       end |> first Union{Missing, Nothing, Int64}
```

This PR:
- enables constant prop' for each union signatures, by calling
  `abstract_call_method_with_const_args` just after each
  `abstract_call_method`
- refactor `abstract_call_method_with_const_args` into two separate
  parts, 1.) heuristics to decide whether to do constant prop', 2.) try
  constant propagation
The added test cases will should showcase the cases where the
inference result could be improved by that.

---

I've not seen notable regression in latency with this PR.
Here is a sample benchmark of the impact of this PR on latency,
from which I guess this PR is acceptable ?

> build time: master (caeacef)
```bash
Sysimage built. Summary:
Total ───────  61.615938 seconds
Base: ───────  26.575732 seconds 43.1313%
Stdlibs: ────  35.038024 seconds 56.8652%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 30/30
Executing precompile statements... 1378/1378
Precompilation complete. Summary:
Total ─────── 116.417013 seconds
Generation ──  81.077365 seconds 69.6439%
Execution ───  35.339648 seconds 30.3561%
    LINK usr/lib/julia/sys.dylib
```
> build time: this PR
```bash
Stdlibs total  ──── 34.077962 seconds
Sysimage built. Summary:
Total ───────  61.804573 seconds
Base: ───────  27.724077 seconds 44.8576%
Stdlibs: ────  34.077962 seconds 55.1383%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 30/30
Executing precompile statements... 1362/1362
Precompilation complete. Summary:
Total ─────── 111.262672 seconds
Generation ──  83.535305 seconds 75.0794%
Execution ───  27.727367 seconds 24.9206%
    LINK usr/lib/julia/sys.dylib
```

> first time to plot: master (caeacef)
```julia
julia> using Plots; @time plot(rand(10,3))
  3.614168 seconds (5.47 M allocations: 324.564 MiB, 5.73% gc time,
53.02% compilation time)
```
> first time to plot: this PR
```julia
julia> using Plots; @time plot(rand(10,3))
  3.557919 seconds (5.53 M allocations: 328.812 MiB, 2.89% gc time,
51.94% compilation time)
```

---

- fixes #37610
- some part of this code was taken from #37637
- this PR is originally supposed to be alternative and more generalized
  version of #39296

aviatesk commented:
Let me flag this as WIP because this PR doesn't use constant analysis results in inlining when there are multiple results.

For now this PR forms `ConstCallInfo` only when there is the one applicable method to pass the test suite and try this PR with nanosoldier benchmark.

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

aviatesk commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1182,25 +1189,32 @@ function analyze_single_call!(ir::IRCode, todo::Vector{Pair{Int, Any}}, idx::Int
         isinvoke::Bool, todo::Vector{Pair{Int, Any}})
-    item = InliningTodo(info.result, sig.atypes, calltype)
-    validate_sparams(item.mi.sparam_vals) || return
+    # TODO effectively use mutliple constant analysis results
+    length(info.results) == 1 || return false
+    result = info.results[1]
+    isa(result, InferenceResult) || return false
```
@Keno do you have any idea of how we can effectively use multiple constant analysis results ?

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/39305#issuecomment-787432874) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/d03508a_vs_c79309b/report.md). cc @christopher-dG

aviatesk commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/39305#issuecomment-788561168) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/b0154e3_vs_5e7aaa6/report.md). cc @christopher-dG

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -42,9 +42,12 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
+        applicable_argtypes = Vector{Any}[] # arrays like `argtypes`, including constants, for each match
         infos = MethodMatchInfo[]
-        for sig_n in splitsigs
+        for j in 1:length(splitsigs)
+            sig_n = splitsigs[j]
```
```suggestion
        for args_n in split_argtypes
            sig_n = argtypes_to_type(args_n)
```

aviatesk replied:
This should be more robust. Thanks.

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -42,9 +42,12 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
     fullmatch = Bool[]
     if splitunions
         splitsigs = switchtupleunion(atype)
```
```suggestion
```

The `unionsplitcost(atype_params)` would also potentially be `unionsplitcost(argtypes)` (I assume no lattice element can become a union after widenconst if it wasn't before)

aviatesk replied:
AFAIU no "constant" lattice element wraps `Union` type, so this `unionsplitcost(argtypes)` should be valid.
I added a note on this. Does this makes sense and do you think it's worth to be left ?
https://github.com/JuliaLang/julia/blob/e5a51a4cb4b19b388c056cd75a42d04b31d05213/base/compiler/abstractinterpretation.jl#L39-L40

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -58,6 +61,10 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
             push!(infos, MethodMatchInfo(matches))
             append!(applicable, matches)
+            for _ in 1:length(matches)
+                push!(applicable_argtypes, split_argtypes[j])
```
```suggestion
                push!(applicable_argtypes, args_n)
```

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -58,6 +61,10 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
             end
             push!(infos, MethodMatchInfo(matches))
             append!(applicable, matches)
+            for _ in 1:length(matches)
```
I think for code readability, these ops are clearer if they go in the same loop
```suggestion
            for m in matches
                push!(applicable, m)
```

aviatesk replied:
I agree. Applied.

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -147,22 +165,23 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
+            # try constant propagation with argtypes for this match
+            # this is in preparation for inlining, or improving the return result
+            this_argtypes = applicable_argtypes === nothing ? argtypes : applicable_argtypes[i]
+            const_this_rt, const_result = abstract_call_method_with_const_args(interp, this_rt, f, this_argtypes, match, sv, edgecycle)
```
`edgecycle` hasn't been properly computed yet, so it cannot be used. But I think you meant `edgecycle1` here, since the other arguments are all call-specific?

aviatesk replied:
Ah, good catch. I overlooked this. Thanks.

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1164,25 +1172,33 @@ function analyze_single_call!(ir::IRCode, todo::Vector{Pair{Int, Any}}, idx::Int
-    item = InliningTodo(info.result, sig.atypes, calltype)
-    validate_sparams(item.mi.sparam_vals) || return
+    # when multiple matches are found, bail out and later inliner will union-split this signature
+    # TODO effectively use mutliple constant analysis results here
```
```suggestion
    # TODO effectively use multiple constant analysis results here
```

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -643,6 +473,197 @@ function abstract_call_method(interp::AbstractInterpreter, method::Method, @nosp
+                     istopfunction(f, :(==)) || istopfunction(f, :!=) ||
+                     istopfunction(f, :<=) || istopfunction(f, :>=) || istopfunction(f, :<) || istopfunction(f, :>) ||
+                     istopfunction(f, :<<) || istopfunction(f, :>>))
+        # it is almost useless to inline the op of when all the same type,
```
```suggestion
        # it is almost useless to inline the op when all the same type,
```

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -643,6 +473,197 @@ function abstract_call_method(interp::AbstractInterpreter, method::Method, @nosp
+end
+
+# This is a heuristic to avoid trying to const prop through complicated functions
+# where we would spend a lot of time, but are probably unliekly to get an improved
```
```suggestion
# where we would spend a lot of time, but are probably unlikely to get an improved
```

vtjnash reviewed (approved):
Makes sense to me, nanosoldier results looked reasonable. Some minor nits

aviatesk committed:
apply suggestions

aviatesk commented:
> nanosoldier results looked reasonable.

Some of the benchmark results show a certain regression:
| ID | time ratio | memory ratio |
|----|------------|--------------|
| `["array", "index", "(\"sumelt\", \"BaseBenchmarks.ArrayBenchmarks.ArrayLSLS{Int32, 2}\")"]` | 1.85 (50%) :x: | 1.00 (1%)  |
| `["array", "index", "(\"sumelt_boundscheck\", \"100000:-1:1\")"]` | 2.00 (50%) :x: | 1.00 (1%)  |

And so I tried to repro that on my machine with something like this:
```julia
# ... BaseBenchmarks.jl definitions

for (A, str) in (arrays_iter..., ranges_iter...)
    if isa(A, ArrayLSLS{Int32, 2})
        @info "sumelt" str
        @btime perf_sumelt($A)
    end

    if A == 100000:-1:1
        @info "sumelt_boundscheck" str
        @btime perf_sumelt_boundscheck($A)
    end
end
```
and I couldn't find any effective performance difference:
> on master
```
┌ Info: sumelt
└   str = "ArrayLSLS{Int32, 2}"
  251.105 μs (0 allocations: 0 bytes)
┌ Info: sumelt_boundscheck
└   str = "100000:-1:1"
  52.773 μs (0 allocations: 0 bytes)
```

> on this PR
```
┌ Info: sumelt
└   str = "ArrayLSLS{Int32, 2}"
  250.922 μs (0 allocations: 0 bytes)
┌ Info: sumelt_boundscheck
└   str = "100000:-1:1"
  55.715 μs (0 allocations: 0 bytes)
```

---

My question is, how we should interpret nanosoldier benchmark results ? Is there any threshold that indicates a real performance regression ?

timholy commented:
`git bisect` identified this PR as the source of #43287
